### DISCUSSION TITLE
Question about Entities and Domain.Shared

### DISCUSSION THREAD

**Discussion by minhhungit** [abpframework/abp#16163] [Q&A]
From this document: https://docs.abp.io/en/abp/latest/Best-Practices/Module-Architecture#domain-layer

```
Domain.Shared package, named as CompanyName.ModuleName.Domain.Shared, that contains constants, enums and other types those can be safely shared with the all layers of the module. This package can also be shared to 3rd-party clients. It can not contain entities, repositories, domain services or any other business objects.
```

And in The clean architecture design, Entity is core, it does not depend on any layer but opposite

I wonder why we can not share Entity to 3rd party, and can we put Entity to Shared layer 

![image](https://user-images.githubusercontent.com/2279508/229536867-07a33843-1846-4fe6-a90e-8c9f8197275b.png)

**Comment by gterdem** [comment]
I think this is more related to Clean Architecture instead of the ABP framework itself but I'll try my best to explain.

In the clean architecture, entities are referred as _business domain entities_ that include the core business rules for your application that should be **valid** for all of your applications. These business domain entities are subject to use cases which can be called _application business rules_. For short, the application layer covers the application rules and the entities cover the core business rules.

ABP uses _Domain.Shared_ layer to house some of the common principles for a flexible development experience. These are 
- **Localization**, where it can be used for the whole solution since not all the layers may be depended on the Domain layer (such as Application.Contracts) directly or indirectly.
- **Constants** are used to prevent magical numbers/strings which again can be used in both DTOs (Application.Contracts) and fluent api configuration (database integration layer).
- **Enums** that can be used in both entities and DTOs since you don't need to map an enum to a DTO.

Adding the Entities to the **Domain.Shared** layer is fundementally **breaks the main idea** of the Clean Architecture (or we can call Domain Driven Design). Since based on Clean Architecture, you should be exposing your entities through the application services.

**Comment by minhhungit** [comment]
thank you for explaining @gterdem 

Firstly I have to say, I'm learning clean architecture and ABP, although that, in my opinion I don't think clean architecture is good enough, it is just theory and that's why it has too many variant.

Just want to discuss more about clean architecture, and base on CA, basically "entities" are core layer 
and other things **depend on it direct or indirect**, entity is just simple c# class (since ABP framework adopts DDD, entities become more complex)

I wonder about hiding entity classes and just expose it in only boundary of domain will leak us to use too many DTO, and in reality when write code with ABP, we have to create many DTO. This leads to use automapper to save time, but automapper also has its drawbacks 

Can you show me what do you think, whether we can reuse entity and reduce using DTOs.
